### DISCUSSION TITLE
I'm about to use strimzi kraft mode for production. Is it stable?

### DISCUSSION THREAD

**Discussion by SteveArseneLee** [strimzi/strimzi-kafka-operator#10039] [Q&A] [answered]
I'm a novice data engineer trying to use apache kafka on kubernetes for commercial use. I'm trying to use Kraft mode itself because I heard it's stable, but I'm wondering if it's also stable when using it through strimzi. I'm not sure because I've seen an article about limitation in other discussions. I need advice from a lot of people. Thank you.

**Comment by scholzj** [comment, marked as answer]
KRaft still has some limitations both in Apache Kafka itself or in Strimzi - for example JBOD storage support, scaling of controller nodes, etc. You should probably read the docs and decide for yourself if it is ready for you or not.

**Reply by SteveArseneLee** [reply]
Today I heard KRaft mode is ready to use in production from Confluent engineer. So I thought it was ok... I can't decide haha
Sorry but what is JBOD storage? I wanna use Kraft if there isn't critical drawbacks. Can you plz let me know where can I find some contents about it? I've searched it for a day but still cannot find decisive content...

**Reply by ppatierno** [reply]
It would be interesting to know if the Confluent engineer was referring to a Confluent commercial offering of Kafka (where they could have some proprietary fixes or additional features) or the upstream one. As Jakub said, the main missing thing in Kafka upstream about KRaft is around the controllers scaling. A KIP was accepted for that and will be fixed in Kafka 3.8.0.
You should decide if not scaling controllers is something acceptable for you.
On the Strimzi side, the usage of JBOD multiple disks in KRaft is missing in the 0.40.0 release but it will be included into the 0.41.0 (because it's already merged on main).

**Reply by scholzj** [reply]
Calling something that is not complete and has limitations because some parts are still being implemented "production ready" will always be highly subjective. Someone might not need the missing features and might be fine with using it in production without them. While someone else might find them important and not having them might be a blocker.

**Reply by scholzj** [reply]
> On the Strimzi side, the usage of JBOD multiple disks in KRaft is missing in the 0.40.0 release but it will be included into the 0.41.0 (because it's already merged on main).

JBOD support in Kraft is early access in Kafka itself =. i.e. not considered production ready. So this is not just Strimzi limitation that will be completely addressed in Strimzi 0.41.0.

**Reply by SteveArseneLee** [reply]
First of all, thank you very much for your response. I have an additional question for the development seniors (I mean you guys) To summarize, I understood that the biggest issue is about JBOD storage support and scaling of controllers. So in production, would it be more stable to use bitnami or apache kafka in KRaft mode on kubernetes without Strimzi? Or would it be better to use Strimzi and zookeeper?

**Reply by scholzj** [reply]
Well, as I said, these limitations are in Kafka in the first place. So if they matter to you, then a KRaft cluster with Bitnami does not help you in any way.

**Reply by SteveArseneLee** [reply]
I apologize for asking the obvious over and over again, and I really appreciate your response.

**Reply by scholzj** [reply]
For the JBOD part, I think the decision might be very easy ... if you for example plan to run in production as a small cluster with 100Gi disks per broker and small throughput, JBOD is not likely to add much value. If you plan to run with 10Ti of disks and very big througput, JBOD might be super important. We have some videos on YouTube covering the JBOD (dis)advantages.

The scaling of the controllers is probably trickier to decide on as it relates to infrastructure you use etc.

**Reply by ppatierno** [reply]
+1 on all Jakub said. Think that Strimzi deploys your cluster which is anyway using Kafka (from the upstream community). So if limits are in Kafka, whatever the solution you use to deploy the cluster, they will stay there. Unless any other form of commercial version of Kafka that could have some stuff already fixed.

**Reply by SteveArseneLee** [reply]
My current k8s cluster is 16 cores & 16GB or less for both master and worker, and I have allocated 500GB or less, so I don't think I'm going to allocate that many resources for kafka as well. So I think I can pass on the JBOD issue as Jakub advised, but I'm still not sure about the controller scaling. I'm not sure how I'm going to orient my production environment with kubernetes at work, as I've just installed and used it monolithically on premise for learning purposes. But I want to make it as stable as possible, so I'm looking into different things.

**Reply by shahsavari-mhd** [reply]
> KRaft still has some limitations both in Apache Kafka itself or in Strimzi - for example JBOD storage support, scaling of controller nodes, etc. You should probably read the docs and decide for yourself if it is ready for you or not.

Thanks for your explanations. but what about ceph storage? when using that, these problem gone, yes? what about production grade of kafka, strimzi, ceph combination?

**Reply by scholzj** [reply]
Kafka, in general, requires block storage. So, if you use Ceph block storage, it should work. Ceph has its own set of issues - depending on your infrastructure you might for example have performance issues. But there is nothing specific regarding KRaft, that is a general thing.

**Reply by shahsavari-mhd** [reply]
please correct me, based on your previous saying, kafka and importantly based on storage is not production ready in usual realworld scenarios? do you have any recommendation?

**Reply by scholzj** [reply]
> please correct me, based on your previous saying, kafka and importantly based on storage is not production ready in usual realworld scenarios? do you have any recommendation?

I'm not sure what you mean with that. I did not say anything like that.

**Reply by shahsavari-mhd** [reply]
Thanks for your help.
